src/condition/lfi_detector.cpp

```diff
@@ -19,93 +17,40 @@ namespace {
 
 constexpr std::size_t min_str_len = 5;
 
-std::vector<std::string_view> match_params(std::string_view path, const ddwaf_object &params)
-{
-    std::stack<const ddwaf_object *> stack;
-    std::vector<std::string_view> strings;
-
-    stack.push(&params);
-
-    while (!stack.empty() && stack.size() <= object_limits::default_max_container_depth) {
-        const ddwaf_object &container = *stack.top();
-        stack.pop();
-
-        for (std::size_t i = 0; i < container.nbEntries; ++i) {
-            const ddwaf_object &child = container.array[i];
-
-            if (child.parameterName != nullptr && child.parameterNameLength >= min_str_len) {
-                const std::string_view key{
-                    child.parameterName, static_cast<std::size_t>(child.parameterNameLength)};
-                if (path.find(key) != std::string_view::npos) {
-                    strings.emplace_back(key);
-                }
-            }
-
-            if (object::is_container(&child)) {
-                if (stack.size() < object_limits::default_max_container_depth) {
-                    stack.push(&child);
-                }
-                continue;
-            }
-
-            if (child.type == DDWAF_OBJ_STRING && child.nbEntries >= min_str_len) {
-                const std::string_view value{
-                    child.stringValue, static_cast<std::size_t>(child.nbEntries)};
-                if (path.find(value) != std::string_view::npos) {
-                    strings.emplace_back(value);
-                }
-            }
-        }
-    }
-
-    return strings;
-}
+using lfi_result = std::optional<std::pair<std::string, std::vector<std::string>>>;
 
-std::vector<std::string_view> split(std::string_view str, char sep)
+lfi_result lfi_impl(std::string_view path, const ddwaf_object &params,
+    const exclusion::object_set_ref &objects_excluded, const object_limits &limits,
+    ddwaf::timer &deadline)
 {
-    std::vector<std::string_view> components;
-
-    std::size_t start = 0;
-    while (start < str.size()) {
-        const std::size_t end = str.find(sep, start);
-
-        if (end == start) {
-            // Ignore zero-sized strings
-            start = end + 1;
+    object::kv_iterator it(&params, {}, objects_excluded, limits);
+    for (; it; ++it) {
+        if (deadline.expired()) {
```

Comment:
You're checking the clock at every iteration? That's super expensive, is it not?

Reply:
No, the clock has a counter and only updates every 16 calls or so.

Reply:
I wouldn't worry too much, we'll profile this later but the calls are quite fast thanks to vdso.